lib/Interpreter/IncrementalCUDADeviceCompiler.h

```diff
@@ -0,0 +1,204 @@
+
+    static constexpr unsigned CxxStdCompiledWith() {
+      // The value of __cplusplus in GCC < 5.0 (e.g. 4.9.3) when
+      // either -std=c++1y or -std=c++14 is specified is 201300L, which fails
+      // the test for C++14 or more (201402L) as previously specified.
+      // I would claim that the check should be relaxed to:
+#if __cplusplus > 201402L
+      return 17;
+#elif __cplusplus > 201103L || (defined(LLVM_ON_WIN32) && _MSC_VER >= 1900)
+      return 14;
+#elif __cplusplus >= 201103L
+      return 11;
+#else
+#error "Unknown __cplusplus version"
+#endif
+    }
+
+    ///\brief Contains the arguments for the cling nvptx and the nvidia
+    /// fatbinary tool. The arguments are static and will set at the constructor
+    /// of IncrementalCUDADeviceCompiler.
+    struct CUDACompilerArgs {
+      std::string cppStdVersion = "-std=c++" + std::to_string(CxxStdCompiledWith());
+      std::string optLevel = "-O0";
+      std::string ptxSmVersion = "--cuda-gpu-arch=sm_20";
+      std::string fatbinSmVersion = "--image=profile=compute_20";
+      ///\brief Argument for the fatbinary tool, which is depend, if the OS is
+      /// 32 bit or 64 bit.
+      std::string fatbinArch = "-32";
+      ///\brief True, if the flag -v is set.
+      bool verbose = false;
+      ///\brief True, if the flag -g is set.
+      bool debug = false;
+      ///\brief A list Arguments, which will passed to the clang nvptx.
+      std::vector<std::string> additionalPtxOpt;
+      ///\brief A list Arguments, which will passed to the fatbinary tool.
+      std::vector<std::string> fatbinaryOpt;
+    };
+
+    CUDACompilerArgs m_CuArgs;
+
+    ///\brief The counter responsible to generate a chain of .cu source files
+    /// and .cu.pch files.
+    unsigned int m_Counter;
+
+    ///\brief Is true if all necessary files have been generated and clang and
+    /// cuda NVIDIA fatbinary are found.
+    bool m_Init;
+
+    ///\brief Path to the folder, where all files will put in. Ordinary the tmp
+    /// folder. Have to end with a separator. Can be empty.
+    std::string m_FilePath;
+    ///\brief Path to the fatbin file, which will used by the CUDACodeGen.
+    std::string m_FatbinFilePath;
+    ///\brief Path to a empty dummy.cu file. The file is necessary to generate
+    /// PTX code from the pch files.
+    std::string m_DummyCUPath;
+    ///\brief Path to the PTX file. Will be reused for every PTX generation.
+    std::string m_PTXFilePath;
+    ///\brief Will be used to generate .cu and .cu.pch files.
+    std::string m_GenericFileName;
+
+    ///\brief Path to the clang++ compiler, which will used to compile the pch
+    /// files and the PTX code. Should be in same folder, as the cling.
+    std::string m_ClangPath;
+    ///\brief Path to the NIVDIA tool fatbinary.
+    std::string m_FatbinaryPath;
+
+    ///\brief Contains information about all include paths.
+    ///
+    std::shared_ptr<clang::HeaderSearchOptions> m_HeaderSearchOptions;
```

Comment:
This isn't a copy, it is a pointer. A copy is not useful, because the header search paths can be extended during runtime.
At running the constructor, the pointer doesn't need to point to a valid object, because the first access is at the `generateFatbinary()` method.

Reply:
This design decision based on the problem, that I have to transform nearly every value* from `CompilerInstance` to `std::string` (see `setCuArgs()`), because I have to pass the values as argument to ExecuteAndWait. My idea is generating the arguments in the constructor to keep the code of `generatePCH()`, `generatePTX()` and `generateFatbinary()` simple and avoid redundant work. So, I don't need the `CompilerInstance` again, except `HeaderSearchOptions`. I think, a reference to `HeaderSearchOptions` is better than `CompilerInstance`, because it shows, which part of `CompilerInstance` I really need. 

\* `CudaGpuBinaryFileNames[0]` is the only one value, which I can use direct.

Reply:
Thanks, makes sense.